### PR TITLE
[BEAM-362] Move aggregator scraping API onto Pipeline, make the support code private

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This moves out aggregators from where they are into various places where they belong. Notes:

 - The runner utility for traversing a pipeline and gathering aggregators is moved out
   to `runners-core`. It could reasonably live in the SDK if it had any other use. It is
   always easier to re-introduce it that to remove it, so I have removed it for now.
 - The Dataflow runner has a new dependency on `runners-core`.
   This is a private implementation detail dependency. But it cannot be shaded until
   we move packages so everything is under `org.apache.beam.runners.core`.
 - The user-facing aggregator stuff is moved out of the `sdk.runners` namespace to
   live alongside `PipelineResult`, of which is it a sub-part. We might consider all
   of these living in a namespace having to do with interacting with a "job", but that
   is future work. For now this is internally consistent. We have a thought/goal that
   the `sdk.runners` namespace can go away.

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @dhalperi note added dependency for Dataflow runner
R: @amitsela I added the `os-maven-plugin` to the Spark runner's pom. We have a transitive dependency that is somewhat badly behaved, for now. Once #539 goes in, it may no longer be necessary; I haven't looked closely. The transitive dependency may also be fixed by upstream upgrades by the time it becomes relevant.

**Comment by amitsela** [comment]
Though I'm not sure what's the issue with Travis. I'll take a look once I'm near a computer.

**Comment by lukecwik** [comment]
Ken, I commented on PR/681 that we can use a jar that contains all os's since only depending on one means that if the client that builds the application is running on an OS which is different than what the runner executes on it will fail to load the native library. Once PR/681 is fixed up, I don't think you'll need to have the maven OS detection plugin.

**Comment by kennknowles** [comment]
@lukecwik yea, I know. Same answer as @dhalperi there. A bit unfortunate/temporary, but unblocks for now.

**Comment by kennknowles** [comment]
@amitsela Travis did pass prior to my opening the PR. Seems there was a Travis timeout/hang in the Mac build. "No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself."

I'll go ahead and rebase to kick Travis and see if it was an infrastructure issue.

**Comment by kennknowles** [comment]
Discussed offline a bit, with these conclusions:

 - To produce the JSON from the `TransformHierarchy` we are actually going to need this stuff in the SDK, but it should be properly private and have a nice API.
 - But we still do need to add the dependency from Dataflow runner to runners-core to eliminate duplication of e.g. `UnboundedReadFromBoundedSource`. But this dependency should be shaded immediately, not left as a future task.

I'll update this PR accordingly.

**Comment by kennknowles** [comment]
Based on comments here, I am going to cherry-pick the os-maven-plugin thing.

**Comment by kennknowles** [comment]
PTAL @dhalperi. It isn't perfect, but the support classes are now package-private. They were previously public and in a namespace we want to remove.

Could move the `getAggregatorSteps` back out, perhaps to some static utility class `Pipelines` or `Aggregators`, but it has to be public somewhere. The support class remains an implementation detail of that one method.

**Comment by lukecwik** [comment]
Ken, I just merged pr/701 so you should be able to drop the OS detection plugin now.

**Comment by kennknowles** [comment]
@lukecwik `os-maven-plugin` is no longer part of this PR.

**Comment by kennknowles** [comment]
Travis failure is the issue in the Mac infrastructure.

**Comment by kennknowles** [comment]
Rebased again - same result.

The Mac build hangs while fetching deps. Note that the pre-merge build of [kennknowles/incubator-beam](https://travis-ci.org/kennknowles/incubator-beam/builds/147243614) passed.

So either it is infrastructural, nondeterministic, or there's some Mac-facing bug in mainline. Or some combination. Note that other PRs eventually get past this by repeatedly retrying. I'd rather not bother with that, but don't want to ignore a flaky issue either. Either way, not related to this change.

**Comment by kennknowles** [comment]
R: -@amitsela (no longer relevant to Spark or particularly interesting in any way)

**Comment by kennknowles** [comment]
@dhalperi every failure for the last few rebases has been timeouts or hangs in Travis, various platforms, most recently all of them. There haven't been any code changes in a while. Jenkins has been happy all along. Do you have any comments about the code itself? I'll keep prodding the infrastructure to see if it passes or uncovers a real issue.

**Comment by dhalperi** [comment]
Sorry -- I've just been busy. Will TAL tomorrow.

**Comment by kennknowles** [comment]
It passed for a second :-)

But it had conflicts. I've rebased and pushed. Crossing fingers...

**Comment by kennknowles** [comment]
R: @tgroh maybe a first pass would be good

File: sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java [hunk 1]
```diff
@@ -518,6 +520,14 @@ private String uniquifyInternal(String namePrefix, String origName) {
   }
 
   /**
+   * Returns a {@link Map} between each {@link Aggregator} in the {@link Pipeline} to the {@link
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java]: s/between/from/ ?

**Comment by dhalperi** [comment]
LGTM, please self-merge.

**Pull request merged**